src/openrct2/management/Research.h

```diff
@@ -19,32 +19,80 @@ struct rct_ride_entry;
 
 struct ResearchItem
 {
-    // Bit 16 (0: scenery entry, 1: ride entry)
     union
     {
         uint32_t rawValue;
         struct
         {
-            uint8_t entryIndex;
+            ObjectEntryIndex entryIndex;
             uint8_t baseRideType;
             uint8_t type; // 0: scenery entry, 1: ride entry
-            uint8_t flags;
         };
     };
+    uint8_t flags;
     uint8_t category;
 
     bool IsNull() const;
+    void SetNull();
     bool Equals(const ResearchItem* otherItem) const;
     bool Exists() const;
     bool IsAlwaysResearched() const;
     rct_string_id GetName() const;
 
     ResearchItem() = default;
-    constexpr ResearchItem(uint32_t _rawValue, int32_t _category)
+    constexpr ResearchItem(uint32_t _rawValue, uint8_t _category, uint8_t _flags)
         : rawValue(_rawValue)
+        , flags(_flags)
         , category(_category)
     {
     }
+    ResearchItem(uint8_t _type, ObjectEntryIndex _entryIndex, uint8_t _baseRideType, uint8_t _category, uint8_t _flags)
+        : entryIndex(_entryIndex)
+        , baseRideType(_baseRideType)
+        , type(_type)
+        , flags(_flags)
+        , category(_category)
+    {
+    }
```

Comment:
I think we should use delegating constructor as most as we can to avoid leaving fields uninitialized. If this can be used here, it would would mean converting this
```cpp
    constexpr ResearchItem(uint32_t _rawValue, uint8_t _category, uint8_t _flags)
        : rawValue(_rawValue)
        , flags(_flags)
        , category(_category)
    {
    }
```` 
to something like
```cpp
    constexpr ResearchItem(uint32_t _rawValue, uint8_t _category, uint8_t _flags)
        : ResearchItem(/* 5 parameters */)
    {
    }
```

Reply:
There are two constructors one for each type of the Union. I think as I right now is as good as you can get.